pydantic/main.py

```diff
@@ -504,13 +522,16 @@ def _get_value(
                     skip_defaults=skip_defaults,
                     include=value_include and value_include.for_element(i),
                     exclude=value_exclude and value_exclude.for_element(i),
+                    as_type=as_type,
                 )
                 for i, v_ in enumerate(v)
                 if (not value_exclude or not value_exclude.is_excluded(i))
                 and (not value_include or value_include.is_included(i))
             )
 
         else:
+            if as_type is not None:
+                return parse_as_type(v, as_type)
```

Comment:
@samuelcolvin Found a use for that function after all! This function now behaves properly when called on its own, and it also now raises a more appropriate error message when conversion fails.

However, the error message has very little context -- in order to add it, we'd probably want to bubble the error up one level to the `_iter` function. I'm not sure how you feel about that, or what error type would be appropriate, or if you'd like to handle it totally differently.

I'll put in a test showing the current error message at some point in the next 24 hours (need to sleep now) to make it easier for you to see what it looks like now.

Reply:
okay.

Reply:
surely this means we're re-parsing the data and we're back to the same position as we were at at the beginning: having to call `Model(**data)` to guarantee data is in the right form?

Reply:
surely this means we're re-parsing the data and we're back to the same position as we were at at the beginning: having to call `Model(**data)` to guarantee data is in the right form?

Reply:
Yes, I think you’re right, though validating submodels of BaseModel will still be faster than the safe approach.

I’ll try to think of a way to tweak this that only parses if the type has changed, but doesn’t come with a substantial performance penalty. If I can’t get the perf difference to be negligible I’ll try to rework it to be a separate API. This might be feasible by having a call that does a “strict cast” prior to the dict generation (which has all of the fancy field inclusion/exclusion options etc).

Reply:
@samuelcolvin `as_type` will only take a non-`None` value if `v` is a `BaseModel` instance of the wrong type, or if the field type was not the same (see the `_requires_casting` function and its usages to confirm). And the `BaseModel` case was already handled above.

So this won't result in reparsing the data unless the field type is a new non-container/model type. So, basically, reparsing should only happen if the field type was overridden in a submodel.